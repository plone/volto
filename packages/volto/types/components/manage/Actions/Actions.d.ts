export default Actions;
declare function Actions(props: any): JSX.Element;
declare namespace Actions {
    namespace propTypes {
        let pathname: any;
    }
}
