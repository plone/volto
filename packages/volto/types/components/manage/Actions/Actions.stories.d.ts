export const Actions: any;
declare namespace _default {
    export let title: string;
    export { Actions as component };
    export let decorators: ((Story: any) => import("react/jsx-runtime").JSX.Element)[];
    export let argTypes: {};
}
export default _default;
