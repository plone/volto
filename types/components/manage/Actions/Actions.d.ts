export default Actions;
declare function Actions(props: any): JSX.Element;
declare namespace Actions {
    namespace propTypes {
        let pathname: PropTypes.Validator<string>;
    }
}
import PropTypes from 'prop-types';
