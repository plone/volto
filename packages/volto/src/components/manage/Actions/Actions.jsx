import { useState } from 'react';
import PropTypes from 'prop-types';
import { useDispatch, useSelector, shallowEqual } from 'react-redux';
import { Link } from 'react-router-dom';
import { Dropdown, Icon } from 'semantic-ui-react';
import { toast } from 'react-toastify';
import { FormattedMessage, defineMessages, useIntl } from 'react-intl';

import { cut, copy, copyContent, moveContent } from '@plone/volto/actions';
import { getBaseUrl } from '@plone/volto/helpers';
import { ContentsRenameModal, Toast } from '@plone/volto/components';

const messages = defineMessages({
  cut: {
    id: 'Cut',
    defaultMessage: 'Cut',
  },
  copy: {
    id: 'Copy',
    defaultMessage: 'Copy',
  },
  paste: {
    id: 'Paste',
    defaultMessage: 'Paste',
  },
  rename: {
    id: 'Rename',
    defaultMessage: 'Rename',
  },
  messageCopied: {
    id: '{title} copied.',
    defaultMessage: '{title} copied.',
  },
  messageCut: {
    id: '{title} cut.',
    defaultMessage: '{title} cut.',
  },
  messageDeleted: {
    id: '{title} has been deleted.',
    defaultMessage: '{title} has been deleted.',
  },
  messagePasted: {
    id: 'Item(s) pasted.',
    defaultMessage: 'Item(s) pasted.',
  },
  success: {
    id: 'Success',
    defaultMessage: 'Success',
  },
});

const Actions = (props) => {
  const intl = useIntl();
  const dispatch = useDispatch();
  const [showRename, setshowRename] = useState(false);
  const actions = useSelector((state) => state.actions.actions, shallowEqual);
  const action = useSelector((state) => state.clipboard.action);
  const source = useSelector((state) => state.clipboard.source, shallowEqual);

  const id = useSelector((state) =>
    state.content.data ? state.content.data.id : '',
  );
  const title = useSelector((state) =>
    state.content.data ? state.content.data.title : '',
  );

  const onRenameOk = () => {
    setshowRename(false);
  };

  const onRenameCancel = () => {
    setshowRename(false);
  };

  const fncut = () => {
    dispatch(cut([getBaseUrl(props.pathname)]));

    toast.success(
      <Toast
        success
        title={intl.formatMessage(messages.success)}
        content={intl.formatMessage(messages.messageCut, {
          title: title,
        })}
      />,
    );
  };

  const fncopy = () => {
    dispatch(copy([getBaseUrl(props.pathname)]));
    toast.success(
      <Toast
        success
        title={intl.formatMessage(messages.success)}
        content={intl.formatMessage(messages.messageCopied, {
          title: title,
        })}
      />,
    );
  };

  const paste = () => {
    if (action === 'copy') {
      dispatch(copyContent(source, getBaseUrl(props.pathname)));
    }
    if (action === 'cut') {
      dispatch(moveContent(source, getBaseUrl(props.pathname)));
    }
    toast.success(
      <Toast
        success
        title={intl.formatMessage(messages.success)}
        content={intl.formatMessage(messages.messagePasted)}
      />,
    );
  };

  const rename = () => {
    setshowRename(true);
  };

  return (
    <Dropdown
      item
      id="toolbar-actions"
      trigger={
        <span>
          <Icon name="lightning" size="big" />{' '}
          <FormattedMessage id="Actions" defaultMessage="Actions" />
        </span>
      }
    >
      <Dropdown.Menu>
        {actions.object_buttons &&
          actions.object_buttons.map((item) => {
            switch (item.id) {
              case 'cut':
                return (
                  <Dropdown.Item
                    key={item.id}
                    icon="cut"
                    text={item.title}
                    onClick={fncut}
                  />
                );
              case 'copy':
                return (
                  <Dropdown.Item
                    key={item.id}
                    icon="copy"
                    text={item.title}
                    onClick={fncopy}
                  />
                );
              case 'paste':
                return (
                  <Dropdown.Item
                    key={item.id}
                    icon="paste"
                    text={item.title}
                    onClick={paste}
                    disabled={action === null}
                  />
                );
              case 'delete':
                return (
                  <Link
                    key={item.id}
                    to={`${props.pathname}/delete`}
                    className="item"
                  >
                    <Icon name="trash" />
                    {item.title}
                  </Link>
                );
              case 'rename':
                return (
                  <Dropdown.Item
                    key={item.id}
                    icon="text cursor"
                    text={item.title}
                    onClick={rename}
                  />
                );
              default:
                return null;
            }
          })}

        <ContentsRenameModal
          open={showRename}
          onCancel={onRenameCancel}
          onOk={onRenameOk}
          items={[
            {
              url: props.pathname,
              title: title,
              id: id,
            },
          ]}
        />
      </Dropdown.Menu>
    </Dropdown>
  );
};

Actions.propTypes = {
  pathname: PropTypes.string.isRequired,
};

export default Actions;
