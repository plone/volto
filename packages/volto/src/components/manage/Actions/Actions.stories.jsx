import { injectIntl } from 'react-intl';
import React from 'react';
import ActionsComponent from './Actions';
import { RealStoreWrapper as Wrapper } from '@plone/volto/storybook';

const IntlActionsComponent = injectIntl(ActionsComponent);

function StoryComponent(args) {
  return (
    <Wrapper
      customStore={{
        actions: {
          actions: {
            document_actions: [],
            object: [
              {
                icon: '',
                id: 'view',
                title: 'View',
              },
              {
                icon: '',
                id: 'edit',
                title: 'Edit',
              },
              {
                icon: '',
                id: 'folderContents',
                title: 'Contents',
              },
              {
                icon: '',
                id: 'history',
                title: 'History',
              },
              {
                icon: '',
                id: 'local_roles',
                title: 'Sharing',
              },
            ],
            object_buttons: [
              {
                icon: '',
                id: 'cut',
                title: 'Cut',
              },
              {
                icon: '',
                id: 'copy',
                title: 'Copy',
              },
              {
                icon: '',
                id: 'delete',
                title: 'Delete',
              },
              {
                icon: '',
                id: 'rename',
                title: 'Rename',
              },
              {
                icon: '',
                id: 'ical_import_enable',
                title: 'Enable icalendar import',
              },
            ],
            portal_tabs: [],
            site_actions: [
              {
                icon: '',
                id: 'sitemap',
                title: 'Site Map',
              },
              {
                icon: '',
                id: 'accessibility',
                title: 'Accessibility',
              },
              {
                icon: '',
                id: 'contact',
                title: 'Contact',
              },
            ],
            user: [
              {
                icon: '',
                id: 'preferences',
                title: 'Preferences',
              },
              {
                icon: '',
                id: 'dashboard',
                title: 'Dashboard',
              },
              {
                icon: '',
                id: 'plone_setup',
                title: 'Site Setup',
              },
              {
                icon: '',
                id: 'logout',
                title: 'Log out',
              },
            ],
          },
        },
        clipboard: {},
        content: { data: { id: 'blog', title: 'Blog' } },
        intl: {
          locale: 'en',
          messages: {},
        },
      }}
    >
      <div id="toolbar" style={{ display: 'none' }} />
      <IntlActionsComponent {...args} pathname="/test" />
    </Wrapper>
  );
}

export const Actions = StoryComponent.bind({});

export default {
  title: 'Public components/Actions',
  component: Actions,
  decorators: [
    (Story) => (
      <div className="ui segment form attached" style={{ width: '400px' }}>
        <Story />
      </div>
    ),
  ],
  argTypes: {},
};
