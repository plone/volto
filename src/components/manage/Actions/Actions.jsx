/**
 * Actions component.
 * @module components/manage/Actions/Actions
 */

import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { compose } from 'redux';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import { Dropdown, Icon } from 'semantic-ui-react';
import { FormattedMessage, defineMessages, injectIntl } from 'react-intl';

import { cut, copy, copyContent, moveContent } from '@plone/volto/actions';
import { getBaseUrl } from '@plone/volto/helpers';
import { ContentsRenameModal, Toast, toast } from '@plone/volto/components';

const messages = defineMessages({
  cut: {
    id: 'Cut',
    defaultMessage: 'Cut',
  },
  copy: {
    id: 'Copy',
    defaultMessage: 'Copy',
  },
  paste: {
    id: 'Paste',
    defaultMessage: 'Paste',
  },
  rename: {
    id: 'Rename',
    defaultMessage: 'Rename',
  },
  messageCopied: {
    id: '{title} copied.',
    defaultMessage: '{title} copied.',
  },
  messageCut: {
    id: '{title} cut.',
    defaultMessage: '{title} cut.',
  },
  messageDeleted: {
    id: '{title} has been deleted.',
    defaultMessage: '{title} has been deleted.',
  },
  messagePasted: {
    id: 'Item(s) pasted.',
    defaultMessage: 'Item(s) pasted.',
  },
  success: {
    id: 'Success',
    defaultMessage: 'Success',
  },
});

/**
 * Actions container class.
 * @class Actions
 * @extends Component
 */
class Actions extends Component {
  /**
   * Property types.
   * @property {Object} propTypes Property types.
   * @static
   */
  static propTypes = {
    actions: PropTypes.shape({
      object: PropTypes.arrayOf(PropTypes.object),
      object_buttons: PropTypes.arrayOf(PropTypes.object),
      user: PropTypes.arrayOf(PropTypes.object),
    }),
    pathname: PropTypes.string.isRequired,
    id: PropTypes.string.isRequired,
    title: PropTypes.string.isRequired,
    action: PropTypes.string,
    source: PropTypes.arrayOf(PropTypes.string),
    cut: PropTypes.func.isRequired,
    copy: PropTypes.func.isRequired,
    copyContent: PropTypes.func.isRequired,
    moveContent: PropTypes.func.isRequired,
  };

  /**
   * Default properties
   * @property {Object} defaultProps Default properties.
   * @static
   */
  static defaultProps = {
    action: null,
    actions: null,
    source: null,
  };

  /**
   * Constructor
   * @method constructor
   * @param {Object} props Component properties
   * @constructs Actions
   */
  constructor(props) {
    super(props);
    this.cut = this.cut.bind(this);
    this.copy = this.copy.bind(this);
    this.paste = this.paste.bind(this);
    this.rename = this.rename.bind(this);
    this.onRenameOk = this.onRenameOk.bind(this);
    this.onRenameCancel = this.onRenameCancel.bind(this);
    this.state = {
      showRename: false,
    };
  }

  /**
   * On rename ok
   * @method onRenameOk
   * @returns {undefined}
   */
  onRenameOk() {
    this.setState({
      showRename: false,
    });
  }

  /**
   * On rename cancel
   * @method onRenameCancel
   * @returns {undefined}
   */
  onRenameCancel() {
    this.setState({
      showRename: false,
    });
  }

  /**
   * Cut handler
   * @method cut
   * @returns {undefined}
   */
  cut() {
    this.props.cut([getBaseUrl(this.props.pathname)]);
    toast.success(
      <Toast
        success
        title={this.props.intl.formatMessage(messages.success)}
        content={this.props.intl.formatMessage(messages.messageCut, {
          title: this.props.title,
        })}
      />,
    );
  }

  /**
   * Copy handler
   * @method copy
   * @returns {undefined}
   */
  copy() {
    this.props.copy([getBaseUrl(this.props.pathname)]);
    toast.success(
      <Toast
        success
        title={this.props.intl.formatMessage(messages.success)}
        content={this.props.intl.formatMessage(messages.messageCopied, {
          title: this.props.title,
        })}
      />,
    );
  }

  /**
   * Paste handler
   * @method paste
   * @returns {undefined}
   */
  paste() {
    if (this.props.action === 'copy') {
      this.props.copyContent(
        this.props.source,
        getBaseUrl(this.props.pathname),
      );
    }
    if (this.props.action === 'cut') {
      this.props.moveContent(
        this.props.source,
        getBaseUrl(this.props.pathname),
      );
    }
    toast.success(
      <Toast
        success
        title={this.props.intl.formatMessage(messages.success)}
        content={this.props.intl.formatMessage(messages.messagePasted)}
      />,
    );
  }

  /**
   * Rename handler
   * @method rename
   * @returns {undefined}
   */
  rename() {
    this.setState({
      showRename: true,
    });
  }

  /**
   * Render method.
   * @method render
   * @returns {string} Markup for the component.
   */
  render() {
    return (
      <Dropdown
        item
        id="toolbar-actions"
        trigger={
          <span>
            <Icon name="lightning" size="big" />{' '}
            <FormattedMessage id="Actions" defaultMessage="Actions" />
          </span>
        }
      >
        <Dropdown.Menu>
          {this.props.actions.object_buttons &&
            this.props.actions.object_buttons.map((item) => {
              switch (item.id) {
                case 'cut':
                  return (
                    <Dropdown.Item
                      key={item.id}
                      icon="cut"
                      text={item.title}
                      onClick={this.cut}
                    />
                  );
                case 'copy':
                  return (
                    <Dropdown.Item
                      key={item.id}
                      icon="copy"
                      text={item.title}
                      onClick={this.copy}
                    />
                  );
                case 'paste':
                  return (
                    <Dropdown.Item
                      key={item.id}
                      icon="paste"
                      text={item.title}
                      onClick={this.paste}
                      disabled={this.props.action === null}
                    />
                  );
                case 'delete':
                  return (
                    <Link
                      key={item.id}
                      to={`${this.props.pathname}/delete`}
                      className="item"
                    >
                      <Icon name="trash" />
                      {item.title}
                    </Link>
                  );
                case 'rename':
                  return (
                    <Dropdown.Item
                      key={item.id}
                      icon="text cursor"
                      text={item.title}
                      onClick={this.rename}
                    />
                  );
                default:
                  return null;
              }
            })}

          <ContentsRenameModal
            open={this.state.showRename}
            onCancel={this.onRenameCancel}
            onOk={this.onRenameOk}
            items={[
              {
                url: this.props.pathname,
                title: this.props.title,
                id: this.props.id,
              },
            ]}
          />
        </Dropdown.Menu>
      </Dropdown>
    );
  }
}

export default compose(
  injectIntl,
  connect(
    (state) => ({
      actions: state.actions.actions,
      action: state.clipboard.action,
      source: state.clipboard.source,
      id: state.content.data ? state.content.data.id : '',
      title: state.content.data ? state.content.data.title : '',
    }),
    {
      cut,
      copy,
      copyContent,
      moveContent,
    },
  ),
)(Actions);
