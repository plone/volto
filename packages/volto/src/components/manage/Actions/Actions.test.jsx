import React from 'react';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-intl-redux';
import { MemoryRouter } from 'react-router-dom';

import Actions from './Actions';

const mockStore = configureStore();

jest.mock('../Contents/ContentsRenameModal', () =>
  jest.fn(() => <div className="RenameModal" />),
);

describe('Actions', () => {
  it('renders an actions component', () => {
    const store = mockStore({
      actions: {
        actions: {
          document_actions: [],
          object: [
            {
              icon: '',
              id: 'view',
              title: 'View',
            },
            {
              icon: '',
              id: 'edit',
              title: 'Edit',
            },
            {
              icon: '',
              id: 'folderContents',
              title: 'Contents',
            },
            {
              icon: '',
              id: 'history',
              title: 'History',
            },
            {
              icon: '',
              id: 'local_roles',
              title: 'Sharing',
            },
          ],
          object_buttons: [
            {
              icon: '',
              id: 'cut',
              title: 'Cut',
            },
            {
              icon: '',
              id: 'copy',
              title: 'Copy',
            },
            {
              icon: '',
              id: 'delete',
              title: 'Delete',
            },
            {
              icon: '',
              id: 'rename',
              title: 'Rename',
            },
            {
              icon: '',
              id: 'ical_import_enable',
              title: 'Enable icalendar import',
            },
          ],
          portal_tabs: [],
          site_actions: [
            {
              icon: '',
              id: 'sitemap',
              title: 'Site Map',
            },
            {
              icon: '',
              id: 'accessibility',
              title: 'Accessibility',
            },
            {
              icon: '',
              id: 'contact',
              title: 'Contact',
            },
          ],
          user: [
            {
              icon: '',
              id: 'preferences',
              title: 'Preferences',
            },
            {
              icon: '',
              id: 'dashboard',
              title: 'Dashboard',
            },
            {
              icon: '',
              id: 'plone_setup',
              title: 'Site Setup',
            },
            {
              icon: '',
              id: 'logout',
              title: 'Log out',
            },
          ],
        },
      },
      clipboard: {},
      content: { data: { id: 'blog', title: 'Blog' } },
      intl: {
        locale: 'en',
        messages: {},
      },
    });
    const component = renderer.create(
      <Provider store={store}>
        <MemoryRouter>
          <Actions pathname="/test" />
        </MemoryRouter>
      </Provider>,
    );
    const json = component.toJSON();
    expect(json).toMatchSnapshot();
  });
});
