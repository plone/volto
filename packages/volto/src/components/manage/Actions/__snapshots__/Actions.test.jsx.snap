// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Actions renders an actions component 1`] = `
<div
  aria-expanded={false}
  className="ui item dropdown"
  id="toolbar-actions"
  onBlur={[Function]}
  onChange={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onMouseDown={[Function]}
  role="listbox"
  tabIndex={0}
>
  <span>
    <i
      aria-hidden="true"
      className="lightning big icon"
      onClick={[Function]}
    />
     
    Actions
  </span>
  <i
    aria-hidden="true"
    className="dropdown icon"
    onClick={[Function]}
  />
  <div
    className="menu transition"
  >
    <div
      className="item"
      onClick={[Function]}
      role="option"
    >
      <i
        aria-hidden="true"
        className="cut icon"
        onClick={[Function]}
      />
      <span
        className="text"
      >
        Cut
      </span>
    </div>
    <div
      className="item"
      onClick={[Function]}
      role="option"
    >
      <i
        aria-hidden="true"
        className="copy icon"
        onClick={[Function]}
      />
      <span
        className="text"
      >
        Copy
      </span>
    </div>
    <a
      className="item"
      href="/test/delete"
      onClick={[Function]}
    >
      <i
        aria-hidden="true"
        className="trash icon"
        onClick={[Function]}
      />
      Delete
    </a>
    <div
      className="item"
      onClick={[Function]}
      role="option"
    >
      <i
        aria-hidden="true"
        className="text cursor icon"
        onClick={[Function]}
      />
      <span
        className="text"
      >
        Rename
      </span>
    </div>
    <div
      className="RenameModal"
    />
  </div>
</div>
`;
